npm/vite-dev-server/src/resolveConfig.ts

```diff
@@ -4,52 +4,84 @@
+import { Cypress, CypressSourcemap } from './plugins/index'
 import type { Vite } from './getVite'
+import { dynamicImport } from './dynamic-import'
 
 const debug = debugFn('cypress:vite-dev-server:resolve-config')
 
 export const createViteDevServerConfig = async (config: ViteDevServerConfig, vite: Vite): Promise<InlineConfig> => {
-  const { specs, cypressConfig, viteConfig: viteOverrides } = config
-  const root = cypressConfig.projectRoot
-  const { default: findUp } = await importModule('find-up')
-  const configFile = await findUp(configFiles, { cwd: root } as { cwd: string })
-
-  // INFO logging, a lot is logged here.
-  // debug('all dev-server options are', options)
-
-  if (configFile) {
-    debug('resolved config file at', configFile, 'using root', root)
-  } else if (viteOverrides) {
-    debug(`Couldn't find a Vite config file, however we received a custom viteConfig`, viteOverrides)
+  const { viteConfig: inlineViteConfig, cypressConfig: { projectRoot } } = config
+  let resolvedOverrides: InlineConfig = {}
+
+  if (inlineViteConfig) {
+    debug(`Received a custom viteConfig`, inlineViteConfig)
+
+    if (typeof inlineViteConfig === 'function') {
+      resolvedOverrides = await inlineViteConfig()
+    } else if (typeof inlineViteConfig === 'object') {
+      resolvedOverrides = inlineViteConfig
+    }
+
+    // Set "configFile: false" to disable auto resolution of <project-root>/vite.config.js
+    resolvedOverrides = { configFile: false, ...resolvedOverrides }
   } else {
-    if (config.onConfigNotFound) {
-      config.onConfigNotFound('vite', root, configFiles)
-      // The config process will be killed from the parent, but we want to early exit so we don't get
-      // any additional errors related to not having a config
-      process.exit(0)
-    } else {
-      throw new Error(`Your component devServer config for vite is missing a required viteConfig property, since we could not automatically detect one.\n Please add one to your ${config.cypressConfig.configFile}`)
+    const { findUp } = await dynamicImport<typeof import('find-up')>('find-up')
+
+    const configFile = await findUp(configFiles, { cwd: projectRoot })
+
+    if (!configFile) {
+      if (config.onConfigNotFound) {
+        config.onConfigNotFound('vite', projectRoot, configFiles)
+        // The config process will be killed from the parent, but we want to early exit so we don't get
+        // any additional errors related to not having a config
+        process.exit(0)
+      } else {
+        throw new Error(`Your component devServer config for vite is missing a required viteConfig property, since we could not automatically detect one.\n Please add one to your ${config.cypressConfig.configFile}`)
+      }
     }
+
+    debug('Resolved config file at', configFile, 'using root', projectRoot)
+
+    resolvedOverrides = { configFile }
   }
 
+  const finalConfig = vite.mergeConfig(
+    resolvedOverrides,
+    makeCypressViteConfig(config, vite),
```

Comment:
This used to be swapped, but we don't want users overriding our base config. This makes it so our vite config takes precedence (same was how we handle Webpack)